content/docs/rfcs/32/README.md

```diff
@@ -0,0 +1,554 @@
+```
+
+For registration, the user needs to submit their `identity_commitment` (along with any additional registration requirements) to the registry. 
+Upon registration, they should receive `leaf_index` value which represents their position in the merkle tree. 
+Receiving a `leaf_index` is not a hard requirement and is application specific. 
+The other way around is the users calculating the `leaf_index` themselves upon successful registration.
+
+## Signaling
+
+After registration, 
+the users can participate in the application by sending signals to the other participants in a decentralised manner or to a centralised server. 
+Along with their signal, 
+they need to generate a ZK-Proof by using the circuit with the specification described above. 
+
+For generating a proof, 
+the users need to obtain the required parameters or compute them themselves, 
+depending on the application implementation and client libraries supported by the application.
+For example the users can store the membership merkle tree on their end and 
+generate a merkle proof whenever they want to generate a signal. 
+
+### Implementation notes
+
+#### Signal hash
+
+The signal hash can be generated by hashing the raw signal (or content) using the `keccak256` hash function.
+
+#### External nullifier
+
+The external nullifier can be generated by hashing a raw string (i.e UNIX timestamp) value using `keccak256`.
+
+####  Obtaining merkle proof
+
+The merkle proof should be obtained locally or from a trusted third party. 
+By using the [incremental merkle tree algorithm](https://github.com/appliedzkp/incrementalquintree/blob/master/ts/IncrementalQuinTree.ts), 
+the merkle can be obtained by providing the `leaf_index` of the `identity_commitment`. 
+The proof (`merkle_proof`) is composed of the following fields:
+
+```
+{
+  root: bigint
+  indices: number[]
+  path_elements: bigint[][]
+}
+```
+
+
+#### Generating proof
+
+For proof generation, 
+the user need to submit the following fields to the circuit:
+
+```
+    {
+      identity_secret: identity_secret_hash,
+      path_elements: merkle_proof.path_elements,
+      identity_path_index: merkle_proof.indices,
+      x: signal_hash,
+      external_nullifier: external_nullifier,
+      rln_identifier: rln_identifier
+    }
+```
+
+
+#### Calculating output
+
+The proof output is calculated locally, 
+in order for the required fields for proof verification to be sent along with the proof. 
+The proof output is composed of the `y` share of the secret equation and the `internal_nullifier`. 
+The `internal_nullifier` represents an unique fingerprint for the user for the `external_nullifier`.
+The following fields are needed for proof output calculation:
+
+```
+{
+  identity_secret_hash: bigint, 
+  external_nullifier: bigint, 
+  rln_identifier: bigint,
+  x: bigint, 
+}
+```
+
+The output `[y, internal_nullifier]` is calculated in the following way:
+
+```
+a_0 = identity_secret
+a_1 = poseidonHash([a0, external_nullifier])
```

Comment:
The role of RLN identifier is to obfuscate the nullifer so that a user generating a signal for two different apps at the same epoch is impossible to identify (impossible without additional extra metadata, the additional metadata being the user's secret).  For this reason the RLN identifier is used in the internal nullifier as a strong enough measure to prevent this issues. There are potentially other solutions for this, I am not 100% sure if this can be embedded in the a1 coefficient without having significant impact on the other parts of the architecture.

Reply:
> that a user generating a signal for two different apps at the same epoch is impossible to identify

One way is to do a brute-force on the published secret shares (of all the peers) for each epoch to recover various secret keys and see if their commitment appears in the membership group.

Reply:
Agreed, the brute force attacks possibilities are open, especially if:
- the same epoch scheme is used across different applications
- an epoch is chosen inappropriately ("larger" epoch increases the probability of messages being sent to different application in the same epoch) 
- the anonymity sets are small

We'll look further on how to resolve this problem, and also reconsider the possibility of brute force attacks.